hbase-asyncfs/src/test/java/org/apache/hadoop/hbase/io/asyncfs/TestFanOutOneBlockAsyncDFSOutput.java

```diff
@@ -272,4 +279,132 @@ public void testWriteLargeChunk() throws IOException, InterruptedException, Exec
     }
     assertArrayEquals(b, actual);
   }
+
+  /**
+   * <pre>
+   * This test is for HBASE-26679. Consider there are two dataNodes: dn1 and dn2,dn2 is a slow DN.
+   * The threads sequence before HBASE-26679 is:
+   * 1.We write some data to {@link FanOutOneBlockAsyncDFSOutput} and then flush it, there are one
+   *   {@link FanOutOneBlockAsyncDFSOutput.Callback} in
+   *   {@link FanOutOneBlockAsyncDFSOutput#waitingAckQueue}.
+   * 2.The ack from dn1 arrives firstly and triggers Netty to invoke
+   *   {@link FanOutOneBlockAsyncDFSOutput#completed} with dn1's channel, then in
+   *   {@link FanOutOneBlockAsyncDFSOutput#completed}, dn1's channel is removed from
+   *   {@link FanOutOneBlockAsyncDFSOutput.Callback#unfinishedReplicas}.
+   * 3.But dn2 responds slowly, before dn2 sending ack,dn1 is shut down or have a exception,
+   *   so {@link FanOutOneBlockAsyncDFSOutput#failed} is triggered by Netty with dn1's channel,
+   *   and because the {@link FanOutOneBlockAsyncDFSOutput.Callback#unfinishedReplicas} does not
+   *   contain dn1's channel,the {@link FanOutOneBlockAsyncDFSOutput.Callback} is skipped in
+   *   {@link FanOutOneBlockAsyncDFSOutput#failed} method,and
+   *   {@link FanOutOneBlockAsyncDFSOutput#state} is set to
+   *   {@link FanOutOneBlockAsyncDFSOutput.State#BROKEN},and dn1,dn2 are all closed at the end of
+   *   {@link FanOutOneBlockAsyncDFSOutput#failed}.
+   * 4.{@link FanOutOneBlockAsyncDFSOutput#failed} is triggered again by dn2 because it is closed,
+   *   but because {@link FanOutOneBlockAsyncDFSOutput#state} is already
+   *   {@link FanOutOneBlockAsyncDFSOutput.State#BROKEN},the whole
+   *   {@link FanOutOneBlockAsyncDFSOutput#failed} is skipped. So wait on the future
+   *   returned by {@link FanOutOneBlockAsyncDFSOutput#flush} would be stuck for ever.
+   * After HBASE-26679, for above step 4,even if the {@link FanOutOneBlockAsyncDFSOutput#state}
+   * is already {@link FanOutOneBlockAsyncDFSOutput.State#BROKEN}, we would still try to trigger
+   * {@link FanOutOneBlockAsyncDFSOutput.Callback#future}.
+   * </pre>
+   */
+  @Test
+  public void testFlushStuckWhenOneDataNodeFailedBeforeOtherDataNodeAck() throws Exception {
```

Comment:
The problem for introducing a separated test class in my patch is that, we need to increase the READER_TIMEOUT value, we set it to 2 seconds in this test but then it will generate unexpected netty traficc and mess up the testing code here. The default value is 60 seconds, which is enough for this test.

So I suggest we introduce a separated test class.

Reply:
@Apache9 , why we need to increase the READER_TIMEOUT value?

Reply:
As said above, it will mess up the testing code. We will send heartbeat to DN when there is no out going packet, the timeout is controlled by READER_TIMEOUT. If we set it to 2 seconds, it will keep sending package out and DN will respond immedately and then mess up the testing handler added by us.

Reply:
@Apache9 , I mean it seems not affect the test? The Response for heartbeat is ignored by your added ChannelInboundHandler? and in AckHandler.channelRead0, it is also ignore.

Reply:
OK, seems no big problem if we just ignore all inbound messages.

But it needs extra consideration, better not introduce more conditions to a test. When I first tried to implement the UT, I used a boolean flag to drop exact one package and the heartbeat message messed up the test. In the future if we want to change the implementation to simulate more, maybe another developer will fall into this too and waste more time...